backend/cmm_helpers.mli

```diff
@@ -1079,15 +1079,29 @@ val unboxed_float32_array_ref :
   expression -> expression -> Debuginfo.t -> expression
 
 (** Read an unboxed float32 from a 64-bit field in an array represented as
-    a mixed block, as used for unboxed product arrays. *)
+    a mixed block (with tag zero), as used for unboxed product arrays.
+
+    The float32 is expected to be in the least significant bits of the
+    64-bit field.
+
+    The zero-indexed element number is specified as a tagged immediate.
+    This always coincides with the index in words.
+*)
 val unboxed_mutable_float32_unboxed_product_array_ref :
-  expression -> expression -> Debuginfo.t -> expression
+  expression -> element_num:expression -> Debuginfo.t -> expression
 
 (** Write an unboxed float32 into a 64-bit field in an array represented as
-    a mixed block, as used for unboxed product arrays. *)
+    a mixed block (with tag zero), as used for unboxed product arrays.
+
+    The zero-indexed element number is specified as a tagged immediate.
+    This always coincides with the index in words.
+
+    The float32 will be written to the least significant bits of the
+    64-bit field.  The top 32 bits of the written word are undefined.
```

Comment:
Just a thought: couldn't this behaviour (leaving the top bits undefined) be annoying/problematic in certain situations ? I'm thinking if such an array ever gets stored inside a cmx file for instance (which I'd find reasonable to store some constants of that type), then cmx files that should be equal may actually be different (and have different hashes/digests), which may confuse build systems ?

Reply:
I think this is okay because static constants will pad the extra space with zeros - or at least that's what float32 scalar constants are doing already.

Reply:
I don't think that actually solves the case I was thinking about: in that situation, let's assume we are compiling some `foo.ml` file, and in that file there is an array of unboxed  `float32` that can be lifted. When compiling `foo.ml`, the compiler will lift the array so that it will be a static constant of `Foo`, but during compilation, it is not a static constant, and is likely to be represented as a non-static array of `float32` (created dynamically from the AST of the source expression). That non-static array will then be stored in the cmx files, as part of an flambda2 static let binding, and so it will be marshalled when creating the cmx files.

In that situation, the cmx file will contain the marshalled version of a non-static array of unboxed `float32`. Since the higher bits are left uninitialized, that means that the cms file would in essence be non-deterministic: re-compiling the same file without any changes would yield different cmx files, since the uninitialized bits would likely change beetween invocations of the compiler.

More generally, any other program that tries and marshall dynamic arrays for unboxed `float32` would have very surprising behaviours (the compiler with cmx files being just one instance), and I'm not sure that's a nice situation to have.

Reply:
Ah, I see. Currently marshalling arrays of unboxed types is not supported, though - do we plan to add it?

Reply:
As long as we don't marshall these, we should be safe indeed. However, with the reinterpret casts on read (and maybe writes ?) to arrays, there will be situations where these uninitialized bits end up being read (and maybe written to ?) though it would be a programming error to write such code.

Reply:
I've decided to make this write function zero the whole word, then write the 32-bit float.  This seems probably more consistent with the behaviour in the int32 case, where the top half has a defined semantics.  The read functions in both cases ignore the top half, though.